sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/UtilsAnnotationVisitor.java

```diff
@@ -0,0 +1,406 @@
+                        statements.add(
+                            new StatementDef.If(
+                                new ExpressionDef.Condition(" == ", isPropertyEqual, ExpressionDef.constant(false)),
+                                ExpressionDef.constant(false).returning()
+                            )
+                        );
+                    }
+                }
+                statements.add(new StatementDef.Return(ExpressionDef.constant(true)));
+                return StatementDef.multi(statements);
+            });
+        classDefBuilder.addMethod(method);
+    }
+
+    /*
+    Creates a hashCode method with signature:
+        public static int BeanNameUtils.hashCode(BeanName object)
+     */
+    private static void createHashCodeMethod(ClassDef.ClassDefBuilder classDefBuilder, String objectName, List<PropertyElement> properties) {
+        MethodDef method = MethodDef.builder("hashCode")
+            .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+            .addParameter("object", ClassTypeDef.of(objectName))
+            .returns(int.class)
+            .build((self, parameterDef) -> {
+                List<StatementDef> statements = new ArrayList<>();
+                ExpressionDef propertyHashExpression;
+                VariableDef.Local hashValue = new VariableDef.Local("hashValue", TypeDef.of(int.class));
+                VariableDef arrays = new VariableDef.Local("java.util.Arrays", ClassTypeDef.of(java.util.Arrays.class));
+
+                // initialize hash value at 1
+                statements.add(new StatementDef.DefineAndAssign(hashValue, ExpressionDef.constant(1)));
+
+                // add all property values to hash calculations
+                TypeDef propertyTypeDef;
+                ExpressionDef thisProperty;
+                for (PropertyElement beanProperty : properties) {
+                    propertyTypeDef = TypeDef.of(beanProperty.getType());
+                    if (beanProperty.getReadMethod().isPresent()) {
+                        thisProperty = parameterDef.get(0).asVariable().invoke(
+                            beanProperty.getReadMethod().get().getSimpleName(),
+                            propertyTypeDef,
+                            List.of()
+                        );
+                    } else {
+                        continue;
+                    }
+
+                    // calculate new property hash value
+                    if (propertyTypeDef instanceof TypeDef.Array) {
+                        String methodName = (((TypeDef.Array) propertyTypeDef).dimensions() > 1) ?  "deepHashCode" : "hashCode";
+                        propertyHashExpression = arrays.invoke(methodName, TypeDef.of(int.class), thisProperty);
+                    } else if (propertyTypeDef instanceof TypeDef.Primitive) {
+                        String typeName = ((TypeDef.Primitive) propertyTypeDef).name();
+                        if (propertyTypeDef == TypeDef.BOOLEAN) {
```

Comment:
rewrite these nested `ifs` as a switch expression that assigns `propertyHashExpression`